tests/cpp/common/test_host_device_vector.cu

```diff
@@ -178,20 +178,66 @@ TEST(HostDeviceVector, TestCopy) {
   SetCudaSetDeviceHandler(nullptr);
 }
 
-TEST(HostDeviceVector, Reshard) {
+TEST(HostDeviceVector, Shard) {
   std::vector<int> h_vec (2345);
   for (size_t i = 0; i < h_vec.size(); ++i) {
     h_vec[i] = i;
   }
   HostDeviceVector<int> vec (h_vec);
   auto devices = GPUSet::Range(0, 1);
 
-  vec.Reshard(devices);
+  vec.Shard(devices);
   ASSERT_EQ(vec.DeviceSize(0), h_vec.size());
   ASSERT_EQ(vec.Size(), h_vec.size());
   auto span = vec.DeviceSpan(0);  // sync to device
 
-  vec.Reshard(GPUSet::Empty());  // pull back to cpu, empty devices.
+  vec.Reshard(GPUDistribution::Empty(), GPUDataPreservation::kPreserve);  // pull back to cpu, empty devices.
+  ASSERT_EQ(vec.Size(), h_vec.size());
+  ASSERT_TRUE(vec.Devices().IsEmpty());
+
+  auto h_vec_1 = vec.HostVector();
+  ASSERT_TRUE(std::equal(h_vec_1.cbegin(), h_vec_1.cend(), h_vec.cbegin()));
+}
+
+TEST(HostDeviceVector, Reshard_PreserveGPUData) {
```

Comment:
A single-device resharding test is actually not very useful. Consider having a simple test for 1 device, and more advanced tests for multiple devices.

Reply:
Done.